src/term/transform/float_combinators.rs

```diff
@@ -1,277 +1,160 @@
-      rule.body.float_combinators(def_name, builtin, &mut combinators);
+      let mut seen = IndexSet::new();
+      rule.body.float_combinators(&mut combinators, &mut name_gen, &slf, def_name, builtin, &mut seen);
     }
 
-    // Definitions are not inserted to the book as they are defined to appease the borrow checker.
-    // Since we are mut borrowing the rules we can't borrow the book to insert at the same time.
     self.defs.extend(combinators);
   }
 }
 
-type Combinators = BTreeMap<Name, Definition>;
-
-#[derive(Debug)]
-struct TermInfo<'d> {
-  // Number of times a Term has been detached from the current Term
-  counter: u32,
-  def_name: Name,
-
-  builtin: bool,
-  needed_names: BTreeSet<Name>,
-  combinators: &'d mut Combinators,
-}
-
-impl<'d> TermInfo<'d> {
-  fn new(def_name: Name, builtin: bool, combinators: &'d mut Combinators) -> Self {
-    Self { counter: 0, def_name, builtin, needed_names: BTreeSet::new(), combinators }
-  }
-  fn request_name(&mut self, name: &Name) {
-    self.needed_names.insert(name.clone());
-  }
-
-  fn provide(&mut self, name: Option<&Name>) {
-    if let Some(name) = name {
-      self.needed_names.remove(name);
-    }
-  }
-
-  fn has_no_free_vars(&self) -> bool {
-    self.needed_names.is_empty()
-  }
-
-  fn replace_scope(&mut self, new_scope: BTreeSet<Name>) -> BTreeSet<Name> {
-    std::mem::replace(&mut self.needed_names, new_scope)
-  }
-
-  fn merge_scope(&mut self, target: BTreeSet<Name>) {
-    self.needed_names.extend(target);
-  }
-
-  fn detach_term(&mut self, term: &mut Term) {
-    let comb_name = Name::new(format!("{}$C{}", self.def_name, self.counter));
-    self.counter += 1;
-
-    let comb_var = Term::Ref { nam: comb_name.clone() };
-    let extracted_term = std::mem::replace(term, comb_var);
+impl Term {
+  fn float_combinators(
+    &mut self,
+    combinators: &mut Combinators,
+    name_gen: &mut usize,
+    book: &Book,
+    def_name: &Name,
+    builtin: bool,
+    seen: &mut IndexSet<Name>,
+  ) {
+    Term::recursive_call(move || {
+      for term in self.children_mut() {
+        // Don't float if it's safe or has unscoped variables.
+        if term.is_safe(book, seen) || term.has_unscoped_diff() {
+          continue;
+        }
 
-    let rules = vec![Rule { body: extracted_term, pats: Vec::new() }];
-    let rule = Definition { name: comb_name.clone(), rules, builtin: self.builtin };
-    self.combinators.insert(comb_name, rule);
-  }
-}
+        // Recursively float the children terms.
+        term.float_combinators(combinators, name_gen, book, def_name, builtin, seen);
 
-#[derive(Debug)]
-enum Detach {
-  /// Can be detached freely
-  Combinator,
-  /// Can not be detached
-  Unscoped { lams: BTreeSet<Name>, vars: BTreeSet<Name> },
-  /// Should be detached to make the program not hang
-  Recursive,
-}
+        match term {
+          // If it is an Application without free variables like '(bar 0)', float into a new definition.
+          Term::App { .. } => {
```

Comment:
We don't need to handle app as an exception. All terms with children could have this logic of first trying to float the children and then trrying to float the term itself.

As an optimization we can consider an application with many arguments to be a single term with many children, as well as considering lambdas to have as children the children of the lambda body.
That would be only to avoid fragmenting the combinators too much, but the overall logic is the same.